folium/plugins/search.py

```diff
@@ -2,83 +2,105 @@
+                collapsed: {{this.collapsed}},
+                textPlaceholder: '{{this.placeholder}}',
             {% if this.geom_type == 'Point' %}
                 initial: false,
-                zoom: {{this.search_zoom}},
+                zoom: {{this.layer.get_name()}}searchZoom?{{this.layer.get_name()}}searchZoom:{{this.parent_map_name}}.getZoom(),
                 position:'{{this.position}}',
                 hideMarkerOnCollapse: true
-            {% endif %}
-            {% if this.geom_type == 'Polygon' %}
+            {% else %}
                 marker: false,
                 moveToLocation: function(latlng, title, map) {
-                var zoom = {{this._parent.get_name()}}.getBoundsZoom(latlng.layer.getBounds());
-                    {{this._parent.get_name()}}.setView(latlng, zoom); // access the zoom
+                var zoom = {{this.layer.get_name()}}searchZoom?{{this.layer.get_name()}}searchZoom:map.getBoundsZoom(latlng.layer.getBounds())
+                    map.setView(latlng, zoom); // access the zoom
                 }
             {% endif %}
                 });
-                searchControl.on('search:locationfound', function(e) {
-
-                    e.layer.setStyle({fillColor: '#3f0', color: '#0f0'});
+                {{this.layer.get_name()}}searchControl.on('search:locationfound', function(e) {
+                    {{this.layer.get_name()}}.setStyle(function(feature){
+                        return feature.properties.style
+                    })
+                    {% if this.options %}
+                    e.layer.setStyle({{ this.options }});
+                    {% endif %}
                     if(e.layer._popup)
                         e.layer.openPopup();
-
-                }).on('search:collapsed', function(e) {
-
-                    {{this.get_name()}}.eachLayer(function(layer) {   //restore feature color
-                        {{this.get_name()}}.resetStyle(layer);
+                })
+                {{this.layer.get_name()}}searchControl.on('search:collapsed', function(e) {
+                        {{this.layer.get_name()}}.setStyle(function(feature){
+                            return feature.properties.style
                     });
                 });
-            {{this._parent.get_name()}}.addControl( searchControl );
+            {{this.parent_map_name}}.addControl( {{this.layer.get_name()}}searchControl );
 
         {% endmacro %}
         """)  # noqa
 
-    def __init__(self, data, search_zoom=12, search_label='name', geom_type='Point', position='topleft'):
+    def __init__(self, layer=None, search_label='name', search_zoom=None, geom_type='Point', position='topleft',
+                 placeholder='Search', collapsed=True, **kwargs):
         super(Search, self).__init__()
-        self.position = position
-        self.data = data
+        assert isinstance(layer, folium.GeoJson), "Search can only be added to GeoJson objects."
+        self.layer = layer
```

Comment:
Is `layer` the best name for the argument? Why not `geojson`?

Reply:
The search plugin can accommodate a wide range of feature layers, not just `L.geoJson`s. Example [here](http://labs.easyblog.it/maps/leaflet-search/examples/simple.html) - the source is embedded down towards the bottom of the page. I think using `layer` as a more general kwarg would be better to open up this functionality to other map layers down the road in future PRs.

Reply:
Looked into it further, and sure enough, the `Search` plugin can accept any type of `FeatureGroup` or `MarkerCluster` as an argument. If it is one of these types of aggregator types instead of `GeoJson`, we just need to omit the `propertyName` in the object that is passed to the constructor, and it will automatically search the indivdual marker for their `title` property. So with `Marker`, `CircleMarker`, etc. classes, we just need to expose a `title` kwarg to allow for this functionality.

I'd like to incorporate this functionality in this PR, but not sure if the `title` kwarg additions to those classes, would be best undertaken in this PR, or a separate one. Let me know what you think.